modules/custom/d_product/d_product.module

```diff
@@ -0,0 +1,193 @@
+<?php
+
+/**
+ * @file
+ * File containing d_product module.
+ */
+
+use Drupal\Core\Link;
+use Drupal\Core\Url;
+use Drupal\node\Entity\Node;
+use Drupal\taxonomy\Entity\Term;
+use Drupal\user\Entity\User;
+
+/**
+ * implement hook_preprocess_HOOK.
+ */
+function d_product_preprocess_node(&$variables) {
+  if (isset($variables['view_mode']) && $variables['view_mode'] == 'teaser') {
+    if (isset($variables['node'])) {
+      $node = $variables['node'];
+      if ($node->getType() == 'd_product') {
+        if (isset($variables['content']['field_product_image'][0])) {
+          $variables['main_image'] = $variables['content']['field_product_image'][0];
+          unset($variables['content']['field_product_image']);
+        }
+
+        if (isset($variables['content']['field_product_categories'][0])) {
+          $variables['category'] = $variables['content']['field_product_categories'];
+          unset($variables['content']['field_product_categories']);
+        }
+
+        if (isset($variables['content']['field_d_product_link'])) {
+          $variables['link'] = $variables['content']['field_d_product_link'];
+          unset($variables['content']['field_d_product_link']);
+        }
+      }
+    }
+  }
+  if (isset($variables['view_mode']) && $variables['view_mode'] == 'full') {
+    $node = $variables['node'];
+    if ($node->hasField('field_d_ask_for_product')) {
```

Comment:
Ja bym to wszystko na starcie ograniczył do rodzaju zawartości d_product,bo to jest już poza IF-em